spread/openstack.go

```diff
@@ -0,0 +1,631 @@
+func errorTitle(msg string) string {
+	node, err := html.Parse(strings.NewReader(msg))
+	if err != nil {
+		return ""
+	}
+	return findErrorTitle(node)
+}
+
+func (p *openstackProvider) findFlavor(flavorName string) (nova.Entity, error) {
+	var flavor nova.Entity
+
+	flavors, err := p.computeClient.ListFlavors()
+	if err != nil {
+		return flavor, fmt.Errorf("failed to retrieve flavors list: %s", errorTitle(err.Error()))
+	}
+
+	for _, f := range flavors {
+		if f.Name == flavorName {
+			flavor = f
+			break
+		}
+	}
+
+	if flavor.Id == "" {
+		return flavor, fmt.Errorf("specified flavor not found: %s", flavorName)
+	}
+
+	return flavor, nil
+}
+
+func (p *openstackProvider) findNetwork(name string) (neutron.NetworkV2, error) {
+	var serverNetwork neutron.NetworkV2
+
+	networks, err := p.networkClient.ListNetworksV2()
+	if err != nil {
+		return serverNetwork, fmt.Errorf("failed to retrieve networks list: %s", errorTitle(err.Error()))
+	}
+
+	// When there are not networks defined, the first network which is not external
+	// is returned (external networks could not be allowed to request)
+	for _, net := range networks {
+		if (name == "" && net.External == false) || (name != "" && name == net.Name) {
+			return net, nil
+		}
+	}
+	return serverNetwork, &FatalError{fmt.Errorf("no valid network found to create floating IP")}
+}
+
+func (p *openstackProvider) findImage(imageName string) (glance.ImageDetail, error) {
+	var sameImage glance.ImageDetail
+	var lastImage glance.ImageDetail
+	var lastCreatedDate time.Time
+
+	images, err := p.imageClient.ListImagesDetail()
+	if err != nil {
+		return sameImage, fmt.Errorf("failed to retrieve images list: %s", errorTitle(err.Error()))
+	}
+
+	for _, i := range images {
+		if i.Name == imageName {
+			sameImage = i
+		} else if strings.Contains(i.Name, imageName) {
+			// Check if the creation date for the current image is after the previous selected one
+			currCreatedDate, err := time.Parse(time.RFC3339, i.Created)
+			// When the creation date is not set or it cannot be parsed, it is considered as created just now
+			if err != nil {
+				currCreatedDate = time.Time{}
+			}
+
+			// Save the image when either it is the first match or it is newer than the previous match
+			if lastImage.Id == "" || currCreatedDate.After(lastCreatedDate) {
+				lastImage = i
+				lastCreatedDate = currCreatedDate
+			}
+		}
+	}
+
+	// return the image when it matchs exactly with the provided name
+	if sameImage.Id != "" {
+		return sameImage, nil
+	}
+
+	if lastImage.Id != "" {
+		return lastImage, nil
+	}
+
+	return sameImage, &FatalError{fmt.Errorf("No matching image found")}
+}
+
+func (p *openstackProvider) findAvailabilityZone() (nova.AvailabilityZone, error) {
+	var zone nova.AvailabilityZone
+
+	zones, err := p.computeClient.ListAvailabilityZones()
+	if err != nil {
+		return zone, fmt.Errorf("failed to retrieve availability zones: %s", errorTitle(err.Error()))
+	}
+
+	if len(zones) == 0 {
```

Comment:
This can be written as:
```go
	if len(zones) == 0 {
		return zone, &FatalError{errors.New("cannot find any availability zones")}
	}
	return zones[0], nil
}
```